src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs

```diff
@@ -413,34 +414,46 @@ public bool TryCreateType(ITypeSymbol symbol, [NotNullWhen(true)] out CSharpType
 
         public bool TryCreateType(ITypeSymbol symbol, Func<System.Type, bool> validator, [NotNullWhen(true)] out CSharpType? type)
         {
-            type = null;
-            INamedTypeSymbol? namedTypeSymbol = symbol as INamedTypeSymbol;
-            if (namedTypeSymbol == null)
+            if (TryCreateTypeForIArrayTypeSymbol(symbol, out type))
+            {
+                return true;
+            }
+
+            // We can only handle IArrayTypeSymbol of framework type and INamedTypeSymbol for now since CSharpType can't represent other types such as IArrayTypeSymbol of user types
+            // Instead of throwing an exception, wihch causes more side effects, we just return false and let the caller handle it.
+            if (symbol is not INamedTypeSymbol namedTypeSymbol)
             {
-                throw new InvalidCastException($"Unexpected type {symbol}");
+                return false;
             }
 
+            return TryCreateTypeForINamedTypeSymbol(symbol, validator, out type, namedTypeSymbol);
+        }
+
+        private bool TryCreateTypeForINamedTypeSymbol(ITypeSymbol symbol, Func<Type, bool> validator, [NotNullWhen(true)] out CSharpType? type, INamedTypeSymbol namedTypeSymbol)
```

Comment:
From where this method is called, the parameter `symbol` and `namedTypeSymbol` is actually the same instance, maybe we could change it like this:
```suggestion
        private bool TryCreateTypeForINamedTypeSymbol(INamedTypeSymbol namedTypeSymbol, Func<Type, bool> validator, [NotNullWhen(true)] out CSharpType type)
```

Reply:
refactoring this with above `switch` change, but `CSharpType?` needs to be nullable

Reply:
OK - if compiler throws the error, maybe `NotNullWhen` attribute works differently from the attribute `MaybeNullWhen` because the latter does not require the question mark (with or without both work fine)

Reply:
Interesting, I would prefer to have the explicit nullable symbol anyway for readability.